weave/workers/weave_scorer.py

```diff
@@ -0,0 +1,445 @@
+    # We want to differentiate between a call not written yet and a call existing but not matching the filter.
+    # - first we count the calls that match the call id
+    # - if that returns 0, we raise an exception in order to trigger a retry
+    # - if that returns > 0, we know the call exists, so we can proceed with the filter
+
+    count_req = tsi.CallsQueryStatsReq(
+        project_id=ended_call.project_id,
+        filter=tsi.CallsFilter(call_ids=[ended_call.id]),
+    )
+
+    count = server.calls_query_stats(count_req).count
+
+    if count == 0:
+        raise CallNotWrittenError(f"Call {ended_call.id} not yet written to DB")
+
+    req = tsi.CallsQueryReq(
+        project_id=ended_call.project_id,
+        filter=tsi.CallsFilter(
+            call_ids=[ended_call.id],
+            query=query,
+            op_names=op_names,
+        ),
+    )
+
+    calls = server.calls_query(req).calls
+
+    if len(calls) == 0:
+        logger.warning("No matching calls found for call id %s", ended_call.id)
+        return None
+
+    if len(calls) > 1:
+        logger.warning("Multiple calls found for call id %s", ended_call.id)
+        return None
+
+    call = calls[0]
+
+    if not call.ended_at:
+        return None
+
+    if call.exception:
+        return None
+
+    logger.info("Found call %s", call.id)
+
+    return build_client_call(call)
+
+
+def build_client_call(server_call: tsi.CallSchema) -> Call:
+    """Converts a server call to a client call."""
+    server = get_trace_server()
+
+    return Call(
+        _op_name=server_call.op_name,
+        project_id=server_call.project_id,
+        trace_id=server_call.trace_id,
+        parent_id=server_call.parent_id,
+        id=server_call.id,
+        inputs=from_json(server_call.inputs, server_call.project_id, server),
+        output=from_json(server_call.output, server_call.project_id, server),
+        exception=server_call.exception,
+        summary=dict(server_call.summary) if server_call.summary is not None else None,
+        _display_name=server_call.display_name,
+        attributes=server_call.attributes,
+        started_at=server_call.started_at,
+        ended_at=server_call.ended_at,
+        deleted_at=server_call.deleted_at,
+    )
+
+
+async def process_monitor(
+    monitor: Monitor,
+    monitor_internal_ref: InternalObjectRef,
+    ended_call: tsi.EndedCallSchemaForInsert,
+    wb_user_id: str,
+) -> None:
+    """Actually apply the monitor's scorers for an ended call."""
+    if (call := get_filtered_call(monitor.op_names, monitor.query, ended_call)) is None:
+        return
+
+    for scorer in monitor.scorers:
+        logger.info("Applying scorer %s to call %s", scorer.name, call.id)
+        await apply_scorer(
+            monitor_internal_ref, scorer, call, ended_call.project_id, wb_user_id
+        )
+
+
+def _do_score_call(scorer: Scorer, call: Call, project_id: str) -> tuple[str, Any]:
```

Comment:
This method (and a few others) are very similar to what is already in the client. Was it possible to reuse that code / refactor it to do so? If not, a very clear comment and explanation would be useful as they will need to be kept in sync.